google-api-client/src/main/java/com/google/api/client/googleapis/services/AbstractGoogleClient.java

```diff
@@ -328,9 +447,34 @@ protected Builder(
         HttpRequestInitializer httpRequestInitializer) {
       this.transport = Preconditions.checkNotNull(transport);
       this.objectParser = objectParser;
-      setRootUrl(rootUrl);
-      setServicePath(servicePath);
+      this.rootUrl = normalizeRootUrl(rootUrl);
+      this.servicePath = normalizeServicePath(servicePath);
       this.httpRequestInitializer = httpRequestInitializer;
+      this.serviceName = parseServiceName(rootUrl);
+      this.isUserConfiguredEndpoint = !defaultEndpointRegex.matcher(this.rootUrl).matches();
```

Comment:
I'm still not convinced that we need this. Considering the following scenarios:
1. `rootUrl` is set to non-GDU, `serviceName` would be parsed to null, hence we would use `rootUrl` as is in `determiningEndpoint()`.
2. `rootUrl` is set to GDU, `isUserConfiguredEndpoint` would be false anyway, and we would re-create the `rootUrl` with `serviceName` and `universeDomain`.

Reply:
This does work, but I think it has misleading logic inside for certain edge cases (i.e. regional endpoints). For example, when using a regional endpoint in speech: `https://eu-speech.googleapis.com` and passing that into the constructor, the logic in this PR would work. But I believe it sets the serviceName as `eu-speech` and builds the endpoint, when it _probably_ shouldn't.

The more I think about this, the more I'm leaning towards fixing the serviceName parsing logic the same way we did in GAPICs. The Apiary generator probably should be the one parsing the discovery doc and generating this value as a constant inside the library. That way we don't have worry about parsing the serviceName based on user params or user set values and the Apiary would have this value by default. I probably should have done that to begin with, but I didn't account for regional endpoints and I thought this way would be fine.